File: ankiconnect-server/src/routes/prompts.ts
import { Router } from 'express';
import * as aiService from '../services/ai.js';
import { getSettings } from '../services/settings.js';

export const promptsRouter = Router();

// POST /api/prompts/preview — render prompts for a given input without calling the LLM
promptsRouter.post('/preview', async (req, res) => {
  const { newMessage, highlightedWords } = req.body as {
    newMessage: string;
    highlightedWords?: string[];
  };

  if (!newMessage) {
    res.status(400).json({ error: 'newMessage is required' });
    return;
  }

  const settings = await getSettings();
  const prompts = aiService.getSystemPrompts(settings.showTransliteration);

  const trimmedMessage = newMessage.trim();
  const isSingleWord = !trimmedMessage.includes(' ') && trimmedMessage.length < 30;
  const hasHighlightedWords = highlightedWords && highlightedWords.length > 0;

  let mode: string;
  let systemPrompt: string;
  let userMessage: string;

  if (hasHighlightedWords) {
    mode = 'focused-words';
    systemPrompt = prompts.focusedWords;
    const rendered = aiService.renderUserTemplate('focusedWords', {
      sentence: newMessage,
      highlightedWords: highlightedWords.join(', '),
    });
    userMessage =
      rendered || `Sentence: ${newMessage}\n\nFocus words: ${highlightedWords.join(', ')}`;
  } else if (isSingleWord) {
    mode = 'single-word';
    systemPrompt = prompts.word;
    const rendered = aiService.renderUserTemplate('word', { word: newMessage });
    userMessage = rendered || newMessage;
  } else {
    mode = 'sentence';
    systemPrompt = prompts.sentence;
    const rendered = aiService.renderUserTemplate('sentence', { sentence: newMessage });
    userMessage = rendered || newMessage;
  }

  // Also show card extraction prompt (used as second LLM call)
  const extractionSystemPrompt = prompts.extractCard;

  res.json({
    mode,
    systemPrompt,
    userMessage,
    extractionSystemPrompt,
  });
});

File: python-server/anki_defs/services/routes/prompts.py
"""Prompts route — preview rendered prompts without calling the LLM."""

from __future__ import annotations

import os
from typing import Any

from bottle import request, response

from .. import ai
from ..settings import get_settings


def register(app: Any) -> None:
    @app.post("/api/prompts/preview")
    def preview() -> dict:
        if os.environ.get("ANKI_DEFS_DEV"):
            ai.reload_prompts()

        body = request.json or {}
        new_message: str = body.get("newMessage", "")
        highlighted_words: list[str] | None = body.get("highlightedWords")
        request_mode: str | None = body.get("mode")

        if not new_message:
            response.status = 400
            return {"error": "newMessage is required"}

        settings = get_settings()
        prompts = ai.get_system_prompts(
            settings.get("showTransliteration", False)
        )

        selection = ai.select_prompt(
            prompts,
            new_message,
            highlighted_words=highlighted_words,
            mode=request_mode,
        )

        return {
            "mode": selection.mode,
            "systemPrompt": selection.system_prompt,
            "userMessage": selection.user_message,
        }
